### DISCUSSION TITLE
How to Setup A Project to Use StrongGrid

### DISCUSSION THREAD

**Discussion by enjoau** [Jericho/StrongGrid#397] [Q&A]
Hi Jericho,

I've just discovered your excellent codebase and was wondering what is the best way to implement this into a Web Service / API project?

What sort of project do you recommend? I've tried to create a new Web API 2.0 project but I'm new to MVC and am not sure how to get such a project to work.

I'm trying to parse a signed Webhook but have issues with the lines

- var events = await parser.ParseSignedEventsWebhookAsync(Request.Body, publicKey, signature, timestamp).ConfigureAwait(false);
- return Ok()

The Request.Body shows an error HttpRequestBase does not contain a definition for 'Body' and Return Ok() says The name Ok does not exist in the current context. I've used the exact same code as in your example so am guessing that perhaps a Web API 2.0 project is not the way to go.

Is is possible to implement this code in an existing WebService we have that uses asmx and SOAP?

Do you have any sample projects I could use for reference or further documentation floating around the web somewhere that I might find useful in setting this up?

Apologies for my ignorance but as I say, I'm rather new to this.

Thanks for your time,

Stuart

**Comment by Jericho** [comment]
No matter which framework you chose (MVC, ASPX, etc.) there are only two requirements to be able to parse webhook messages received from SendGrid using StrongGrid:

1. You need to be able to read the so called "payload" POST'ed in the body of the request by SendGrid. Each framework has their own unique way to do this, you just need to figure out what the syntax is for the framework you chose. In the case of MVC for example, you can access the `Request.Body`.
2. You need to be able to send a response back to SendGrid indicating that you acknowledge receipt of the the message they sent to you.  To quote the SendGrid documentation: `your application must return a 200 status code to signal that the message data has been received`. Again, each framework has their own unique way to accomplish this and you'll need to figure out the appropriate syntax that allows you to return a HTTP 200 response. In the case of MVC, this is done by returning an object implementing `IAsyncResult` such as Ok() (see [here](https://docs.microsoft.com/en-us/aspnet/core/web-api/action-return-types?view=aspnetcore-5.0) for more information).

I included the following sample in the readme which I hope provides a good indication: 
```csharp
namespace WebApplication1.Controllers
{
	[Route("api/SendGridWebhooks")]
	public class SendGridController : Controller
	{
		[HttpPost]
		[Route("Events")]
		public async Task<IActionResult> ReceiveEvents()
		{
			var parser = new WebhookParser();
			var events = await parser.ParseWebhookEventsAsync(Request.Body).ConfigureAwait(false);
			
			... do something with the events ...

			return Ok();
		}
	}
}
```

> The Request.Body shows an error HttpRequestBase does not contain a definition for 'Body' and Return Ok() says The name Ok does not exist in the current context

It's hard for me to guess what might cause this error. Maybe you're not using the latest Web API???? Here is what I see in Visual Studio when I create a new web api. Do you see the same?

![image](https://user-images.githubusercontent.com/112710/119200939-f202eb80-ba5b-11eb-9442-81f7d1a7ec1e.png)

The sample I provided works fine if you click on the option I highlight in the screenshot.

> Is is possible to implement this code in an existing WebService we have that uses asmx and SOAP?

As I said, you can use any framework you prefer, you'll simply need to figure out the proper syntax to get the payload and the way to return the respond HTTP 200.

**Reply by enjoau** [reply]
Thanks very much for the detailed answer.
That cleared things up nicely and I have more than enough to continue, just a case of do I use our existing project or get my backside into gear and finally start using .NET Core :)
Thanks again and loving your work
